2.9BSD/usr/man/cat1/mt.1

MT(1)               UNIX Programmer's Manual                MT(1)

NAME
     mt - magnetic tape manipulating program

SYNOPSIS
     mt [ -t _t_a_p_e_n_a_m_e ] _c_o_m_m_a_n_d [ _c_o_u_n_t ]
     mt [ -f _t_a_p_e_n_a_m_e ] _c_o_m_m_a_n_d [ _c_o_u_n_t ]

DESCRIPTION
     _M_t is used to give commands to nine-track tape.  If a tape
     name is not specified, the environment variable TAPE is
     used;  if no such environment variable exists, _m_t uses the
     device /_d_e_v/_n_r_m_t_1.  Note that _t_a_p_e_n_a_m_e must reference a raw
     (not block) tape device.  If the optional _c_o_u_n_t is not
     specified, _m_t uses a count of 1.

     A command can be abbreviated to as few letters as are neces-
     sary to uniquely identify it. Here are the commands and the
     types of tape drives they can apply to:

     Command             Description
     eof, weof           write _c_o_u_n_t end-of-file marks
     fsf                 space forward _c_o_u_n_t files
     fsr                 space forward _c_o_u_n_t records
     bsf                 space backward _c_o_u_n_t files
     bsr                 space backward _c_o_u_n_t records
     rewind              rewind tape
     offline, rewoffl    rewind tape and go offline
     status              return status of the tape unit

FILES
     /dev/rmt*           Raw magnetic tape interface

SEE ALSO
     dd(1), ioctl(2), mt(4), environ(5)

MESSAGES
     usage: mt [ -f device ] command [ count ]
               This is a memory jogger message to tell you how to
               use the _m_t command in the event that you typed
               something wrong.

     mt: don't grok _c_o_m_m_a_n_d
               This message indicates that the magnetic tape _c_o_m_-
               _m_a_n_d that you typed as an argument to _m_t was not
               in the set of requests it recognizes.  See the
               list of correct commands above.

     mt: negative repeat count
               The repeat count for commands such as fsf must be
               positive.

     _c_o_m_m_a_n_d _n_n_n _d_e_v_i_c_e failed: _s_o_m_e _e_r_r_o_r _m_e_s_s_a_g_e

Printed 7/25/83                                                 1

MT(1)               UNIX Programmer's Manual                MT(1)

               The magnetic tape _c_o_m_m_a_n_d applied to _d_e_v_i_c_e with a
               repeat count of _n_n_n did not work for some reason.
               The reason appears as the string
               _s_o_m_e _e_r_r_o_r _m_e_s_s_a_g_e.  Check that the device name
               was correct, or that the command is applicable to
               that device.  For example, the message might
               appear as:

               fsf 1 f/dev/rmt0 failed: File not found

     mt: _s_o_m_e _e_r_r_o_r _m_e_s_s_a_g_e
               This message usually means that you are applying
               the _m_t command to the wrong kind of device, that
               is, a device that the _m_t command does not recog-
               nize.  The _s_o_m_e _e_r_r_o_r _m_e_s_s_a_g_e is often cryptic,
               for example:

               % mt  -f /dev/console  fsf 5

               /dev/console: Inappropriate ioctl for device

     mt: can't open _t_a_p_e_n_a_m_e: _s_o_m_e _e_r_r_o_r _m_e_s_s_a_g_e
               The _m_t command cannot access the device specified
               by _t_a_p_e_n_a_m_e.

     mt: can't get status from _t_a_p_e_n_a_m_e: _s_o_m_e _e_r_r_o_r
               The device specified by _t_a_p_e_n_a_m_e doesn't support
               the system call necessary to obtain the status of
               the tape drive.

EXIT CODES
     0    Successful completion

     1    _M_t did not recognize the command;  the device could not
          be opened.

     2    The _i_o_c_t_l(2) system call to the tape driver failed.

BUGS
     Commands are not interruptible.

Printed 7/25/83                                                 2



2.9BSD/usr/src/ucb/mt.c
#ifndef	lint
static	char *sccsid = "@(#)mt.c	4.8 (Berkeley) 83/05/08";
#endif

/*
 * mt --
 *   magnetic tape manipulation program
 */
#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/mtio.h>
#include <sys/ioctl.h>

#define	equal(s1,s2)	(strcmp(s1, s2) == 0)

struct commands {
	char *c_name;
	int c_code;
	int c_ronly;
} com[] = {
	{ "weof",	MTWEOF,	0 },
	{ "eof",	MTWEOF,	0 },
	{ "fsf",	MTFSF,	1 },
	{ "bsf",	MTBSF,	1 },
	{ "fsr",	MTFSR,	1 },
	{ "bsr",	MTBSR,	1 },
	{ "rewind",	MTREW,	1 },
	{ "offline",	MTOFFL,	1 },
	{ "rewoffl",	MTOFFL,	1 },
	{ "status",	MTNOP,	1 },
	{ 0 }
};

int mtfd;
struct mtop mt_com;
struct mtget mt_status;
char *tape;

main(argc, argv)
	char **argv;
{
	char line[80], *getenv();
	register char *cp;
	register struct commands *comp;

	if (argc > 2 && (equal(argv[1], "-t") || equal(argv[1], "-f"))) {
		argc -= 2;
		tape = argv[2];
		argv += 2;
	} else
		if ((tape = getenv("TAPE")) == NULL)
			tape = DEFTAPE;
	if (argc < 2) {
		fprintf(stderr, "usage: mt [ -f device ] command [ count ]\n");
		exit(1);
	}
	cp = argv[1];
	for (comp = com; comp->c_name != NULL; comp++)
		if (strncmp(cp, comp->c_name, strlen(cp)) == 0)
			break;
	if (comp->c_name == NULL) {
		fprintf(stderr, "mt: don't grok \"%s\"\n", cp);
		exit(1);
	}
	if ((mtfd = open(tape, comp->c_ronly ? 0 : 2)) < 0) {
		perror(tape);
		exit(1);
	}
	if (comp->c_code != MTNOP) {
		mt_com.mt_op = comp->c_code;
		mt_com.mt_count = (argc > 2 ? atoi(argv[2]) : 1);
		if (mt_com.mt_count < 0) {
			fprintf(stderr, "mt: negative repeat count\n");
			exit(1);
		}
		if (ioctl(mtfd, MTIOCTOP, &mt_com) < 0) {
			fprintf(stderr, "%s %s %d ", tape, comp->c_name,
				mt_com.mt_count);
			perror("failed");
			exit(2);
		}
	} else {
		if (ioctl(mtfd, MTIOCGET, (char *)&mt_status) < 0) {
			perror("mt");
			exit(2);
		}
		status(&mt_status);
	}
}

#ifdef vax
#include <vaxmba/mtreg.h>
#include <vaxmba/htreg.h>

#include <vaxuba/utreg.h>
#include <vaxuba/tmreg.h>
#undef b_repcnt		/* argh */
#include <vaxuba/tsreg.h>
#endif

#ifdef sun
#include <sundev/tmreg.h>
#include <sundev/arreg.h>
#endif

#ifdef pdp11
#include <sys/htreg.h>
#include <sys/tmreg.h>
#undef b_repcnt		/* argh */
#include <sys/tsreg.h>
#endif

struct tape_desc {
	short	t_type;		/* type of magtape device */
	char	*t_name;	/* printing name */
	char	*t_dsbits;	/* "drive status" register */
	char	*t_erbits;	/* "error" register */
} tapes[] = {
#ifdef vax
	{ MT_ISTS,	"ts11",		0,		TSXS0_BITS },
	{ MT_ISHT,	"tm03",		HTDS_BITS,	HTER_BITS },
	{ MT_ISTM,	"tm11",		0,		TMER_BITS },
	{ MT_ISMT,	"tu78",		MTDS_BITS,	0 },
	{ MT_ISUT,	"tu45",		UTDS_BITS,	UTER_BITS },
#endif
#ifdef sun
	{ MT_ISCPC,	"TapeMaster",	TMS_BITS,	0 },
	{ MT_ISAR,	"Archive",	ARCH_CTRL_BITS,	ARCH_BITS },
#endif
#ifdef pdp11
	{ MT_ISTS,	"ts11",		0,		TSXS0_BITS },
	{ MT_ISHT,	"tm03",		HTFS_BITS,	HTER_BITS },
	{ MT_ISTM,	"tm11",		0,		TMER_BITS },
#endif
	{ 0 }
};

/*
 * Interpret the status buffer returned
 */
status(bp)
	register struct mtget *bp;
{
	register struct tape_desc *mt;

	for (mt = tapes; mt->t_type; mt++)
		if (mt->t_type == bp->mt_type)
			break;
	if (mt->t_type == 0) {
		printf("unknown tape drive type (%d)\n", bp->mt_type);
		return;
	}
	printf("%s tape drive, residual=%d\n", mt->t_name, bp->mt_resid);
	printreg("ds", bp->mt_dsreg, mt->t_dsbits);
	printreg("\ner", bp->mt_erreg, mt->t_erbits);
	putchar('\n');
}

/*
 * Print a register a la the %b format of the kernel's printf
 */
printreg(s, v, bits)
	char *s;
	register char *bits;
	register unsigned short v;
{
	register int i, any = 0;
	register char c;

	printf("%s=%o", s, v);
	bits++;
	if (v && bits) {
		putchar('<');
		while (i = *bits++) {
			if (v & (1 << (i-1))) {
				if (any)
					putchar(',');
				any = 1;
				for (; (c = *bits) > 32; bits++)
					putchar(c);
			} else
				for (; *bits > 32; bits++)
					;
		}
		putchar('>');
	}
}
